datasync/chaindatafetcher/chaindata_fetcher.go

```diff
@@ -107,9 +127,102 @@ func (f *ChainDataFetcher) Stop() error {
 	logger.Info("wait for all goroutines to be terminated...")
 	f.wg.Wait()
 	logger.Info("terminated all goroutines for chaindatafetcher")
+	f.started = false
+	return nil
+}
+
+func (f *ChainDataFetcher) start() error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.started {
+		return errors.New("the chaindata fetcher is already started")
+	}
+
+	f.chainSub = f.blockchain.SubscribeChainEvent(f.chainCh)
+	// TODO-ChainDataFetcher add logic to request from checkpoint to the first posted event after start.
+
+	f.started = true
 	return nil
 }
 
+func (f *ChainDataFetcher) stop() error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if !f.started {
+		return errors.New("the chaindata fetcher is already stopped")
+	}
+
+	f.chainSub.Unsubscribe()
+	f.started = false
+	return nil
+}
+
+func (f *ChainDataFetcher) startRange(start, end int64, reqType requestType) error {
+	f.rangeMu.Lock()
+	defer f.rangeMu.Unlock()
+	if f.rangeStarted {
+		return errors.New("the chaindata fetcher is already started with range")
+	}
+	f.rangeStarted = true
+	defer func() { f.rangeStarted = false }()
+
+	events, err := f.makeChainEvents(start, end)
+	if err != nil {
+		return err
+	}
+
+	for _, e := range events {
+		f.reqCh <- &request{
+			reqType: requestTypeTransaction,
+			event:   e,
+		}
+		// TODO-ChainDataFetcher add stop logic while processing the events.
+	}
+	return nil
```

Comment:
This won't be returned unless the whole request is finished.

Reply:
We can add a scheduler for Range job later.

Reply:
For now, I divided the making chain event one by one. 9442e828f8e13bcb92540579f2cdc432308a8aa7